services/data_sources/json_notification_data_source.rb

```diff
@@ -0,0 +1,159 @@
+require "securerandom"
+require_relative "notification_data_source"
+require_relative "../../shared/logging_module"
+require_relative "../../shared/json_convertible"
+require_relative "../../shared/models/notification"
+
+module FastlaneCI
+  # Mixin the JSONConvertible class for Notification
+  class Notification
+    include FastlaneCI::JSONConvertible
+  end
+
+  # Data source for notifications backed by JSON
+  class JSONNotificationDataSource < NotificationDataSource
+    include FastlaneCI::JSONDataSource
+    include FastlaneCI::Logging
+
+    class << self
+      attr_accessor :file_semaphore
+    end
+
+    # @return [FastlaneCI::GitRepo]
+    attr_accessor :git_repo
+
+    # Can't have us reading and writing to a file at the same time
+    JSONNotificationDataSource.file_semaphore = Mutex.new
+
+    # Reloads notifications from the notifications data source after instantiation
+    #
+    # @param  [Any] params
+    def after_creation(**params)
+      if params.nil?
+        raise "Either user or a provider credential is mandatory."
+      else
+        if !params[:user] && !params[:provider_credential]
+          raise "Either user or a provider credential is mandatory."
+        else
+          params[:provider_credential] ||= params[:user].provider_credential(type: ProviderCredential::PROVIDER_CREDENTIAL_TYPES[:github])
+          @git_repo = FastlaneCI::GitRepo.new(git_config: self.json_folder_path, provider_credential: params[:provider_credential])
```

Comment:
Thinking about this, I don't think we even want to store this in git.
Here's why, I'm going through the use cases of why we would want persisted notifications:
1. Server restarts
1. User doesn't acknowledge notifications and logs out

So that's great, we just need persistence to handle these cases. The reasons why we want to store it in git?
1. Server crashes and we want to replace it

We haven't really discussed all the places where we want notifications, but the most common ones I can think of:
1. something wrong with tooling (Xcode, whatever)
1. Bot's access to git is messed up
1. Build failed
1. Yey, something good happened (success message)

All of these are either specific to the build machine or, will be triggered again in the future so a new notification will be created. 

So, tldr; I don't think there is any utility in storing our notifications in git. We should just keep it locally somewhere and be ok with losing the notifications.

Reply:
👍 Especially for the MVP, this should be good enough